### ISSUE TITLE
Local variables declared as reference accept anything

### ISSUE THREAD

**Issue by tethal** [qorelanguage/qore#176]
_Code_:

```
reference x;
printf("%N\n", x.type());
x = 4;
printf("%N\n", x.type());
x = "abc";
printf("%N\n", x.type());
```

_Output:_

```
"nothing"
"integer"
"string"
```

_Expected:_
    Error on both lines with assignments.

**Comment by pavelkveton** [comment]
after discussion with Ondra, it seems that the best solution would be to change "reference" from a data type to a keyword that would be allowed only on the very left side (well, as a declaration).

For now, we need something much simplier. The assumptions are:
1) the reference variable is expected to be used as lvalue (otherwise, why not use normal variable? - well, there is a reason, the same one as in C++: the reference can point to a rather complex class, which we want to use as rvalue only, but we don't want to copy it to a local variable).

2) there is no way to change the object the reference points to - the assignment to the reference variable simply means the assignment to the object it refers to, and not the change of the referred object.

3) assignment into "reference to nothing" should always fail (see x=4; above) - like dereferencing the null pointer.

4) using uninitialized reference variable (nothing) as r-value is disputable and requires further review. Conservative way is to forbid it (like in 3).

Given the points 2 and 3, the reference only makes sense if it is initialized immediately at the time of the declaration, like 

int i;
reference x = \i;

So, the bug will be fixed by forbidding the use of uninitialized reference.

Note that the following code throws an exception too:
int i;  
int j;
reference x;
x = \i;
x = \j;
unhandled QORE System exception thrown in TID 1 at 2015-12-10 15:55:28.973392 Thu -01:00 (GMT+1) at <stdin>:5
RUNTIME-TYPE-ERROR: <lvalue> expects type 'int', but got type 'runtime reference to lvalue' instead

If we forbid the use of uninitialized reference, we need (depending on what is the current implementation) either allow for initializing the reference with first (x = \i) construction, or allow for such assignment only with declaration (i.e. reference x = \i;).

**Comment by davidnich** [comment]
keep in mind that Qore also has the "*reference" type - meaning that the lvalue or return value only accepts a reference or no value

I would normally assume that the use of uninitialized / unassigned references would be enforced in the same way as any other unassigned variable.

**Comment by pavelkveton** [comment]
since we have the discussion on the types ongoing and reference will be definitely covered too, I'm putting this bug on hold.

Anyway, I don't understand David's comment "I would normally assume that the use of uninitialized / unassigned references would be enforced in the same way as any other unassigned variable."

Other unassigned variables in Qore have some "default values" (or NOTHING). NOTHING could be used as "default value" for uninitialized references when used as rvalue.

But what is the semantics of uninitialized reference used as lvalue? IMHO it should always be a runtime exception, right? (if so, then we could probably fix the bug for now in this way regardless on the ongoing "types" discussion).

**Comment by davidnich** [comment]
well actually objects, closures & code references cannot have a default value, so in these cases I would expect that we would treat any use of such unassigned lvalues as an error & references the same way

**Comment by davidnich** [comment]
fixed with #1814 and #1815
